app/src/js/components/Form/Form.js

```diff
@@ -42,147 +76,139 @@ const errorMessage = (errors) => `Please review ${errors.join(', ')} and submit
-      if (!markedDirty) {
-        return;
-      }
+    if (this.isInflight()) return;
+
+    this.setState(createNextState((draftState) => {
+      // validate input values in the store
+      // if values pass validation, write to payload object
+      const errors = [];
+      const payload = {};
+      const { inputs, dirty } = draftState;
 
-      // if expected type is json, validate as json first
-      if (input.type === formTypes.textArea && input.mode === 'json') {
-        try {
-          value = JSON.parse(value);
-        } catch (e) {
-          errors.push(input.schemaProperty);
-          return set(inputState, [inputId, 'error'], t.errors.json);
+      this.props.inputMeta.forEach(field => {
+        let inputId = generateComponentId(field.schemaProperty, this.id);
+        let { value } = inputs[inputId];
+
+        // don't set a value for values that haven't changed
+        if (!dirty[inputId]) return;
+
+        // if expected type is json, validate as json first
+        if (field.type === formTypes.textArea && field.mode === 'json') {
+          try {
+            value = JSON.parse(value);
+          } catch (e) {
+            errors.push(field.schemaProperty);
+            inputs[inputId].error = t.errors.json;
+          }
+        } else if (field.type === formTypes.number) {
+          try {
+            value = parseInt(value);
+          } catch (e) {
+            errors.push(field.schemaProperty);
+            inputs[inputId].error = t.errors.integerRequired;
+          }
         }
-      }
 
-      if (input.type === formTypes.number) {
-        try {
-          value = parseInt(value);
-        } catch (e) {
-          errors.push(input.schemaProperty);
-          return set(inputState, [inputId, 'error'], t.errors.integerRequired);
+        if (field.validate && !field.validate(value)) {
+          errors.push(field.schemaProperty);
+          let error = field.error || t.errors.generic;
+          inputs[inputId].error = error;
+        } else if (inputs[inputId].error) {
+          delete inputs[inputId].error;
         }
-      }
 
-      if (input.validate && !input.validate(value)) {
-        errors.push(input.schemaProperty);
-        let error = input.error || t.errors.generic;
-        return set(inputState, [inputId, 'error'], error);
-      } else if (inputState[inputId].error) {
-        delete inputState[inputId].error;
-      }
+        // Ignore empty fields that aren't required
+        // These may have input elements in the form, but
+        // the API will fail if it's sent empty strings
+        if (value !== '' || field.required) {
+          // Use 'set' function since the value of 'input.schemaProperty' may
+          // be a "path", such as 'rule.type', so we would want to set a nested
+          // property (i.e., the 'type' property within a 'rule' property).
+          set(payload, field.schemaProperty, value);
+        }
+      });
```

Comment:
General form comment so not necessarily something we should worry about here, but something's up with required fields. I would expect the first check the form does is that all required fields are filled. If I submit a completely empty form, there's no user feedback. Then, if I fill out the form, only some of the required fields will show the required error vs while others will have an error from the API. We should probably add a ticket to deal with the form logic/error handling/validation better

Reply:
Agreed. There are actually a number of situations where I've seen errors occur silently without indication to the user that something went wrong. Matt and I also discussed this the other day, so I'll create a ticket for it.